Public/Src/FrontEnd/MsBuild/PipConstructor.cs

```diff
@@ -271,16 +297,32 @@ public bool TrySchedulePipForFile(ProjectWithPredictions project, QualifierId qu
                 references = project.ProjectReferences.Where(projectReference => !projectReference.PredictedTargetsToExecute.TargetsAreKnownToBeEmpty);
             }
 
+            var argumentsBuilder = processBuilder.ArgumentsBuilder;
             foreach (ProjectWithPredictions projectReference in references)
             {
-                bool outputsPresent = m_processOutputsPerProject.TryGetValue(projectReference, out ProcessOutputs processOutputs);
-                Contract.Assert(outputsPresent, $"Pips must have been presented in dependency order: {projectReference.FullPath.ToString(PathTable)} missing, dependency of {project.FullPath.ToString(PathTable)}");
+                bool outputsPresent = m_processOutputsPerProject.TryGetValue(projectReference, out MSBuildProjectOutputs projectOutputs);
+                if (!outputsPresent)
+                {
+                    Contract.Assert(false, $"Pips must have been presented in dependency order: {projectReference.FullPath.ToString(PathTable)} missing, dependency of {project.FullPath.ToString(PathTable)}");
+                }
 
-                // Add all known opaque directories
-                foreach (StaticDirectory output in processOutputs.GetOutputDirectories())
+                // Add all known output directories
+                foreach (StaticDirectory output in projectOutputs.OutputDirectories)
                 {
                     processBuilder.AddInputDirectory(output.Root);
                 }
+
+                // If the dependency was built in isolation, this project needs to access the generated cache files
+                if (projectOutputs.BuildsInIsolation)
+                {
+                    var outputCache = projectOutputs.OutputCacheFile;
+                    processBuilder.AddInputFile(outputCache);
+                    using (argumentsBuilder.StartFragment(PipDataFragmentEscaping.NoEscaping, string.Empty))
+                    {
+                        argumentsBuilder.Add(PipDataAtom.FromString("/irc:"));

```

Comment:
/irc is an MSBuild param? Should add a comment what it is for, maybe link to a doc page or the doc comments in code to make it easier to cross-reference

Reply:
Perhaps it would be also clearer since it is programatically generated to use the fully expanded commandline argument

---
In reply to: [271992530](https://github.com/Microsoft/BuildXL/pull/80#discussion_r271992530) [](ancestors = 271992530)

Reply:
Yes, brand new! :) This is part of the new 'build in isolation' mode, available in the latest MSBuild. Not sure if there are docs yet, @cdmihai ? Happy to link to that, otherwise I can just add a regular comment, summarizing what it does

Reply:
Yup `MSBuild /?` has documentation on them. There's also the build in isolation section in the spec: https://github.com/Microsoft/msbuild/blob/master/documentation/specs/static-graph.md#isolated-builds

Reply:
But I suggest you just use the long name of the parameters